src/frontend/src/optimizer/rule/index_selection.rs

```diff
@@ -92,37 +100,175 @@ impl Rule for IndexSelectionRule {
+                            y.index + index.index_item.len(),
+                            index.primary_table.columns[y.index].data_type().clone(),
+                        ))),
+                    ],
+                    DataType::Boolean,
+                )))
+            })
+            .chain(new_predicate.into_iter())
+            .collect_vec();
+        let on = Condition { conjunctions };
+        let join = LogicalJoin::new(
+            index_scan.into(),
+            primary_table_scan.into(),
+            JoinType::Inner,
+            on,
+        );
+
+        // 2. push down predicate, so we can calculate the cost of index lookup
+        let join_ref = join.predicate_pushdown(Condition::true_cond());
+
+        let join_with_predicate_push_down =
+            join_ref.as_logical_join().expect("must be a logical join");
+        let new_join_left = join_with_predicate_push_down.left();
+        let index_scan_with_predicate: &LogicalScan = new_join_left
+            .as_logical_scan()
+            .expect("must be a logical scan");
+
+        // 3. calculate the cost
+        let mut index_table_scan_io_estimator =
+            TableScanIoEstimator::new(index_scan_with_predicate);
+        let index_cost =
+            index_table_scan_io_estimator.estimate(index_scan_with_predicate.predicate());
+        // lookup cost = index cost * LOOKUP_COST_CONST
+        let lookup_cost = index_cost.mul(&IndexCost::new(LOOKUP_COST_CONST));
+
+        // 4. keep the same schema with original logical_scan
+        let scan_output_col_idx = logical_scan.output_col_idx();
+        let lookup_join = join_ref.prune_col(
+            &scan_output_col_idx
+                .iter()
+                .map(|&col_idx| col_idx + offset)
+                .collect_vec(),
+        );
+
+        (lookup_join, lookup_cost)
+    }
+}
+
 struct TableScanIoEstimator<'a> {
     table_scan: &'a LogicalScan,
     row_size: usize,
 }
 
 impl<'a> TableScanIoEstimator<'a> {
     pub fn new(table_scan: &'a LogicalScan) -> Self {
+        // 5 for table_id + 1 for vnode + 8 for epoch
+        let row_meta_field_estimate_size = 14_usize;
+        let table_desc = table_scan.table_desc();
         Self {
             table_scan,
-            row_size: table_scan
-                .table_desc()
-                .columns
-                .iter()
-                .map(|x| TableScanIoEstimator::estimate_data_type_size(&x.data_type))
-                .reduce(|x, y| x + y)
-                .unwrap(),
+            row_size: row_meta_field_estimate_size
+                + table_desc
+                    .columns
+                    .iter()
+                    // add order key twice for its appearance both in key and value
+                    .chain(
+                        table_desc
+                            .order_key
+                            .iter()
+                            .map(|x| &table_desc.columns[x.column_idx]),
+                    )
```

Comment:
We may remember to update this after we dedup the order key in the value. cc @yuhao-su, @wcy-fdu